### PR TITLE
Generate metrics from model's layers

### PR DESCRIPTION
**Pull request by flaviabeo**

### CODE REVIEW

File: scripts/generate_layers_metrics.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+        print(len(prompts))
+        ids = ids.to("cuda")
+    
+    if hasattr(model.config, "ntk_scaling") and model.config.ntk_scaling:
+        max_seq_len = max(max_len, model.config.max_expected_seq_len)
+    else:
+        # without ntk scaling, extending the seq length too far gives bogus results.
+        max_seq_len = model.config.max_expected_seq_len
+
+    result = generate(
```
Comment (kaoutar55) [inline scripts/generate_layers_metrics.py:102]: I think it is better to use model.forward() instead of generate to:
- Trigger one full forward pass without sampling or token iteration.
- See all intermediate activations, since hooks will fire exactly once per layer.
- Avoid introducing noise from sampling, past key caching, etc. 

If we use generate only may mask individual issues inside specific layers because It’s a high-level API that wraps many operations: (forward pass, KV cache logic, sampling or greeting decoding, post-processing). It may skip certain branches inside model.forward() depending on decoding logic (e.g., only decoder path, or only first token). It may use optimized inference paths (e.g., with contiguous_cache=True) that bypass certain logic like residual addition, attention masking, or past key handling.


@jjhursey is this the way that you typically compare between the models at a macro level. Only relying on generate? 
Let us know what you think.

Comment (flaviabeo) [reply scripts/generate_layers_metrics.py]: Thank you! Added the `model.forward` mode

File: scripts/generate_layers_metrics.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    for hook in hooks:
+        hook.remove()
+
+    pt_compile_model_time = time.time() - pt_compile_model_time
+    logger.info(f"PT compile complete, took {pt_compile_model_time:.3f}s")
+
+    return layer_stack
+
+def write_csv(l, path, metric, gpu_layer_shape, cpu_layer_shape, output_shape):
```
Comment (kaoutar55) [inline scripts/generate_layers_metrics.py:244]: The variable name `l` is ambiguous, non-descriptive, and visually similar to the numeral 1. can you replace it with something like layer_output_list or similar for clarity.

File: scripts/generate_layers_metrics.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    with open(path, 'w') as f:
+        f.write(f'{metric}\n')
+        f.write(f'GPU shape {gpu_layer_shape} CPU shape {cpu_layer_shape}\n')
+        f.write(f'Metric shape {output_shape}\n')
+        if not type(l) is float:
```
Comment (kaoutar55) [inline scripts/generate_layers_metrics.py:249]: better to use `if not isinstance(l, float):` instead of `if not type(l) is float:`

File: tests/resources/get_thresholds.py [hunk 4]
```diff
@@ -31,22 +39,98 @@
...
-        for metric_file in metric_files:
-
-            with open(metric_file, "r") as file:
-                next(file)
-                for line in file:
-                    metric_list.append(float(line))
-        print(f"found {len(metric_files)} metric files")
-        print(model, metric, np.percentile(metric_list, 99.0))
+
+        if not layer_mode:
```
Comment (kaoutar55) [inline tests/resources/get_thresholds.py:87]: You're reading and parsing the same CSV files in two separate places based on the layer_mode flag. Duplicated code is harder to maintain and debug. If you want to change how metric files are parsed (e.g. handle missing values, logging), you must do it in two places.
It would be better to refactor this by abstract the CSV loading logic into a helper function. example below. The code is not tested but illustrates the logic. 

for example: 
```python
def load_metric_file(file_path):
    values = []
    try:
        with open(file_path, "r") as f:
            next(f)  # skip header
            for line in f:
                values.append(float(line))
    except StopIteration:
        logger.warning(f"{file_path} was empty or malformed.")
    return values
```
and then replace the two locations with 

```python
if not layer_mode:
    for metric_file in metric_files:
        metric_list.extend(load_metric_file(metric_file))
else:
    for metric_file in metric_files:
        values = load_metric_file(metric_file)
        ...


```

**Review by kaoutar55 — COMMENTED** [review]
Thanks @flaviabeo for all the changes and improvements. The PR lgtm overall. I have some minor changes added and few suggestions:

- We need to add test coverage to tests/ for generate_layers_metrics.py, at least one dry-run test with a toy model. w\We could use a small stub model (2-3 layers) to validate that hooks fire, CSVs are created, and metrics computed.
- For large models (e.g., Granite 8B), layer-wise inference can take time. I suggest to incorporated using tqdm for better UX when generating metrics or reading multiple files.

**Review by kaoutar55 — APPROVED** [review]
Thanks @flaviabeo!

File: scripts/generate_layers_metrics.py [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+common_shapes = list(
+    itertools.product(
+        common_model_paths,
+        common_batch_sizes,
+        common_seq_lengths,
+        common_max_new_tokens,
+    )
+)
+
+def __prepare_inputs(batch_size, seq_length, tokenizer, seed=0):
```
Comment (JRosenkranz) [inline scripts/generate_layers_metrics.py:82]: I think we re-use this in a couple of areas, might make sense to add this to the utils

File: scripts/generate_layers_metrics.py [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+                                             tokenizer=tokenizer)
+
+    assert len(layer_stack_cuda) == len(layer_stack_cpu)
+
+    for layer, cuda_output in layer_stack_cuda:
+        tensor_cpu_out = None
+        tensor_cuda_out = None
+        cos = nn.CosineSimilarity(dim=-1)
+        for cpu_layer, cpu_output in layer_stack_cpu:
+            if cpu_layer == layer:
```
Comment (JRosenkranz) [inline scripts/generate_layers_metrics.py:346]: I think it might make more sense to have each of these lists in a dict instead of a tuple of name/output (given name will be unique). In that case you could iterate over the keys and avoid doing 2 for loops here.

File: scripts/generate_layers_metrics.py [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+                logger.info("CPU Layer {} GPU Layer {}".format(cpu_layer, layer))
+
+                if type(cpu_output) is tuple and type(cuda_output) is tuple:
+                    cos_sim = []
+                    abs_diff = []
+                    if len(cpu_layer) < 2 and len(cpu_layer[-1]) == 1:
```
Comment (JRosenkranz) [inline scripts/generate_layers_metrics.py:352]: can you add a comment for this conditional

File: tests/LAYERS.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+## 1. Generate Metrics by Layer
+
+The idea is to run, the prompts through the model with the pre- and post-hooks added, and then get the metrics for the outputs intercepted by each layer, as in this diagram. Then we can have a baseline with CPU/GPU for a failure threshold in AIU tests. Same idea as the [test_decoders.py](https://github.com/foundation-model-stack/aiu-fms-testing-utils/blob/main/tests/models/test_decoders.py), but for each layer. This way we can measure the discrepancies for the outputs and use the thresholds for detailed debugging problems in AIU.
+
+![metrics generation by layer](./resources/assets/metrics_generation_layers.png)
+
+The script [generate_layers_metrics.py](../scripts/generate_layers_metrics.py) requires the following environment variables:
+
+```bash
+export MODEL_PATHS=ibm-granite/granite-3.2-8b-instruct
```
Comment (JRosenkranz) [inline tests/LAYERS.md:22]: I think it would be better for these to be part of ArgParser

File: scripts/generate_layers_metrics.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+    
+    if hasattr(model.config, "ntk_scaling") and model.config.ntk_scaling:
+        max_seq_len = max(max_len, model.config.max_expected_seq_len)
+    else:
+        # without ntk scaling, extending the seq length too far gives bogus results.
+        max_seq_len = model.config.max_expected_seq_len
+
+    if "generate" in mode:
+        with torch.no_grad():
+            result = generate(
```
Comment (JRosenkranz) [inline scripts/generate_layers_metrics.py:132]: Is this saving all of the layers per iteration of generate?

File: tests/LAYERS.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+Layer model.base_model.layers39.attn.in_proj.key avg cos_sim = 0.9991455078125
+Layer model.base_model.layers39.attn.in_proj.value avg cos_sim = 0.9986572265625
+Layer model.base_model.layers39.attn.in_proj avg cos_sim = 0.0
+Layer model.base_model.layers39.attn.dense avg cos_sim = 0.9987258911132812
+```
+Also, a JSON file is saved to the same output dir. A sample file can be found at: [sample_layer_th.json](https://github.com/flaviabeo/aiu-fms-testing-utils/blob/generate_metrics_layers/tests/resources/sample_layer_th.json)
+
+## 3. Apply the thresholds where needed
+
+In case of AIU debugging tools, the thresholds will be applied to compare AIU outputs with CPU, and then assert if the differences are within the thresholds generated. Bellow, is an architecture of the full integration:
```
Comment (JRosenkranz) [inline tests/LAYERS.md:107]: typo: Bellow -> below

File: scripts/generate_layers_metrics.py (hunk 9 above)
Comment (flaviabeo) [reply scripts/generate_layers_metrics.py:141]: The way I wrote it, it is saving by layer name, so the first iteration completed gets their output measured and saved in csv files. Do you think we should save by iteration as well?

File: scripts/generate_layers_metrics.py (hunk 6 above)
Comment (flaviabeo) [reply scripts/generate_layers_metrics.py:355]: Updated the layer stack structure to dict, it looks way better. Thank you!

File: scripts/generate_layers_metrics.py (hunk 9 above)
Comment (JRosenkranz) [reply scripts/generate_layers_metrics.py:141]: Yes, for this, I'd think we would want to save per iteration of generate, otherwise we are only saving prefill which is essentially the same as just running the forward case

File: aiu_fms_testing_utils/utils/__init__.py [hunk 11]
```diff
@@ -166,3 +167,19 @@ def sample_squad_v2_qa_requests(
         prompt_length_max,
         seed,
     )
+
+def prepare_inputs(batch_size, seq_length, tokenizer, sharegpt_path, seed=0):
```
Comment (JRosenkranz) [inline aiu_fms_testing_utils/utils/__init__.py:171]: can we add documentation for this? Also, is it possible to make the sample_requests as configurable (as we have other sample requests methods)?

Comment (flaviabeo) [reply aiu_fms_testing_utils/utils/__init__.py]: I made the changes to the other files to use the utils' method in this other PR https://github.com/foundation-model-stack/aiu-fms-testing-utils/pull/77. I thought it would be better to not mix these changes in here, then once it's merged I can rebase the other PR. Is this ok?

**Review by JRosenkranz — APPROVED** [review]
lgtm, we can iterate on this and add more features in later PRs
